dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/RelationAnalysis.java

```diff
@@ -863,6 +900,61 @@ public Knowledge visitSameAddress(Relation rel) {
             }
             return new Knowledge(may, enableMustSets ? must : EMPTY_SET);
         }
+
+        @Override
+        public Knowledge visitSameScope(Relation rel, ArrayList<String> scopes) {
+            Set<Tuple> must = new HashSet<>();
+            List<Load> loadEvents = program.getEvents(Load.class);
+            List<Store> storeEvents = program.getEvents(Store.class);
+            List<Fence> fenceEvents = program.getEvents(Fence.class);
+            List<Event> events = new ArrayList<>();
+            events.addAll(loadEvents);
+            events.addAll(storeEvents);
+            events.addAll(fenceEvents);
+            for (Event e1 : events) {
+                for (Event e2 : events) {
+                    if (mustSameScope(e1, e2, scopes) && !exec.areMutuallyExclusive(e1, e2)) {
+                        must.add(new Tuple(e1, e2));
+                    }
+                }
+            }
+            return new Knowledge(must, new HashSet<>(must));
+        }
+
+        @Override
+        public Knowledge visitAlias(Relation rel) {
+            Set<Tuple> must = new HashSet<>();
+            List<MemEvent> events = program.getEvents(MemEvent.class);
+            for (MemEvent e1 : events) {
+                for (MemEvent e2 : events) {
+                    if (directAlias(e1, e2) && !exec.areMutuallyExclusive(e1, e2)) {
+                        must.add(new Tuple(e1, e2));
+                    }
+                }
+            }
+            return new Knowledge(must, new HashSet<>(must));
+        }
+
+        @Override
+        public Knowledge visitSameSpecificScope(Relation rel, ArrayList<String> scopes, String specificScope) {
```

Comment:
I don't understand what the meaning of this relation is

Reply:
I tried to make it a general version for the relation "same-cta" of PTX. Based on their ISCA'2022 paper section 5.4, "The proxy fence must also be inserted in the same CTA where the non-generic access is taking place. Intuitively, a CTA cannot synchronize a different SM’s special-purpose caching." The reference relation in Alloy is [same_block_r](https://github.com/NVlabs/mixedproxy/blob/main/alloy/ptx.als#L182)

Reply:
Classes `SameScope` and `SameSpecificationScope` are very similar. The same is true for methods `visitSameScope` and `visitSameSpecificationScope` (they only differ in one using `mustSameScope` and the other `sameHigherScope`). Those last two methods have also some similarities. I have the impression that all this code can be refactored in a much better way

Reply:
Merged SameSpecificScope into SameScope.